pkg/diagnostics/client/config_contexts.go

```diff
@@ -0,0 +1,289 @@
+with a secure HTTPS connection, the server did not speak HTTPS back to
+us. The most common explanation is that the server listening at that
+port is not the secure server you expected - it may be a non-secure
+HTTP server or the wrong service may be listening there, or you may have
+specified an incorrect port.
+
+You will not be able to connect or do anything at all with the server
+until this server problem is resolved or you specify a corrected
+server address.`
+	clientUnauthn = `
+This means that when we tried to make a request to the master API
+server, your kubeconfig did not present valid credentials to
+authenticate your client. Credentials generally consist of a client
+key/certificate or an access token. Your kubeconfig may not have
+presented any, or they may be invalid.`
+	clientUnauthz = `
+This means that when we tried to make a request to the master API
+server, the request required credentials that were not presented.
+This can happen when an authentication token expires. Try logging in
+with this user again.`
+)
+
+var (
+	invalidCertNameRx = regexp.MustCompile("x509: certificate is valid for (\\S+, )+not (\\S+)")
+)
+
+func (d ConfigContext) Name() string {
+	return fmt.Sprintf("%s[%s]", ConfigContextsName, d.ContextName)
+}
+
+func (d ConfigContext) Description() string {
+	return "Validate client config context is complete and has connectivity"
+}
+
+func (d ConfigContext) CanRun() (bool, error) {
+	if d.RawConfig == nil {
+		// TODO make prettier?
+		return false, errors.New("There is no client config file")
+	}
+
+	if len(d.ContextName) == 0 {
+		return false, errors.New("There is no current context")
+	}
+
+	return true, nil
+}
+
+func (d ConfigContext) Check() types.DiagnosticResult {
+	r := types.NewDiagnosticResult(ConfigContextsName)
+
+	isDefaultContext := d.RawConfig.CurrentContext == d.ContextName
+
+	// prepare bad news message
+	errorKey := "clientCfgError"
+	unusableLine := fmt.Sprintf("The client config context '%s' is unusable", d.ContextName)
+	if isDefaultContext {
+		errorKey = "currentccError"
+		unusableLine = fmt.Sprintf("The current client config context '%s' is unusable", d.ContextName)
+	}
+
+	// check that the context and its constitutuents are defined in the kubeconfig
+	context, exists := d.RawConfig.Contexts[d.ContextName]
+	if !exists {
+		r.Errorf(errorKey, nil, "%s:\n Client config context '%s' is not defined.", unusableLine, d.ContextName)
+		return r
+	}
+	clusterName := context.Cluster
+	cluster, exists := d.RawConfig.Clusters[clusterName]
+	if !exists {
+		r.Errorf(errorKey, nil, "%s:\n Client config context '%s' has a cluster '%s' which is not defined.", unusableLine, d.ContextName, clusterName)
+		return r
+	}
+	authName := context.AuthInfo
+	if _, exists := d.RawConfig.AuthInfos[authName]; !exists {
+		r.Errorf(errorKey, nil, "%s:\n Client config context '%s' has a user identity '%s' which is not defined.", unusableLine, d.ContextName, authName)
```

Comment:
the name of the auth info stanza is not likely to be the user identity... it's typically a fully-qualified username-servername string